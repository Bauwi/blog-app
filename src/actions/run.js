import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

// manage CRUD of run
export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

// set Run posts. This action is dispatched in the header component
// as it needs to be available for the run dropdown

export const runsHasErrored = bool => ({
  type: 'RUN_HAS_ERRORED',
  hasErrored: bool
});

export const runIsLoading = bool => ({
  type: 'RUN_IS_LOADING',
  isLoading: bool
});

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = resume => (dispatch, getState) => {
  dispatch(runIsLoading(true));
  const { uid } = getState().auth;
  const run = [];
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      return run;
    })
    .then(run => dispatch(setRunPosts(run)))
    .then(() => {
      const firstUnreadIndex = run.findIndex(post => post.state === 'unread');
      if (firstUnreadIndex === -1) {
        return dispatch(setCurrentPostRun(run[0].content.id));
      }
      return dispatch(setCurrentPostRun(run[firstUnreadIndex].content.id));
    })
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

// change the status of the past from read to unread
export const updateRunPostToAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_ALREADY_READ',
  id
});

export const startUpdateRunPostToAlreadyRead = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .update({ state: 'read' })
    .then(() => {
      dispatch(updateRunPostToAlreadyRead(id));
    });
};

// this one is not used anymore. But it could be used to mark some popst as unread yet
export const updateRunPostToNotAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ',
  id
});

// Set which post should actually be current. Run navigation depends on this.
// instead of routing, component is updated with currentPost data.
// easier way to deal with "Resume" to provide the user the first unread post in its run.
export const setCurrentPostRun = id => ({
  type: 'SET_CURRENT_POST_RUN',
  id
});

export const startSetCurrentPostRun = id => (dispatch) => {
  dispatch(runIsLoading(true));
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, 10);
  })
    .then(() => dispatch(setCurrentPostRun(id)))
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

export const resetRun = () => ({
  type: 'RESET_RUN'
});

// remove all posts in tun
export const startResetRun = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        db
          .collection('runs')
          .doc(uid)
          .collection('UserRuns')
          .doc(doc.ref.id)
          .delete();
      });
    })
    .then(() => dispatch(resetRun()));
};

// Remove already read posts in run
export const cleanRun = () => ({
  type: 'CLEAN_RUN'
});

export const startCleanRun = alreadyRead => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapShot) => {
      querySnapShot.forEach((doc) => {
        if (alreadyRead.includes(doc.ref.id)) {
          db
            .collection('runs')
            .doc(uid)
            .collection('UserRuns')
            .doc(doc.ref.id)
            .delete();
        }
      });
    })
    .then(dispatch(cleanRun()));
};
