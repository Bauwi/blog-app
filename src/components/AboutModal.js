import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Modal } from 'antd';

export class AboutModal extends Component {
  state = {
    visible: false,
    confirmLoading: false
  };
  showModal = () => {
    this.setState({
      visible: true
    });
  };

  handleCancel = () => {
    this.setState({
      visible: false
    });
  };
  render() {
    const { visible, confirmLoading } = this.state;
    return (
      <div className="login__modal">
        <div onClick={this.showModal}>
          <i className="fa fa-question" /> About
        </div>
        <Modal
          title="About"
          visible={visible}
          confirmLoading={confirmLoading}
          onCancel={this.handleCancel}
          footer={[]}
        >
          <div className="footer__modal">
            <p>This website's single purpose is to talk about art.</p>
            <br />
            <p>This website is a learning project.</p>
            <br />
            <p>
              All rights of articles are reserved to their author, who are credited at the beginning
              of each article.
            </p>
            <br />
            <p>
              If you are an author and want one of the article to be removed, please contact us via
              the contact page.
            </p>
          </div>
        </Modal>
      </div>
    );
  }
}

export default AboutModal;
