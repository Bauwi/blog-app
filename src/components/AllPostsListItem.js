import React from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';

const AllPostsListItem = ({
  id, title, createdAt, shortBody, keywords, authorId
}) => (
  <div>
    <Link to={`${authorId}/read/${id}`}>
      <header>
        <h3>{title}</h3>
        <h5>Published {moment(createdAt).format('MMM Do, YYYY')}</h5>
      </header>
      <p>{shortBody}</p>
      <p>Keywords: {keywords}</p>
    </Link>
  </div>
);

export default AllPostsListItem;
