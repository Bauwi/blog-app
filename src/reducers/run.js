const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
