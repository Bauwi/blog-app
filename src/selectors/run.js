const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
