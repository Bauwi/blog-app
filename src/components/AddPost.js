import React, { Component } from 'react';
import PostForm from './PostForm';
import { connect } from 'react-redux';
import { startAddPost, startAddStar } from '../actions/posts';

export class AddPost extends Component {
  onSubmit = post => {
    this.props.startAddPost(post);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <h1>New Post</h1>
        <PostForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => dispatch(startAddPost(post)),
  startAddStar: id => dispatch(startAddStar(id))
});

export default connect(undefined, mapDispatchToProps)(AddPost);
