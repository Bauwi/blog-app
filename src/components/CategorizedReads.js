import React, { Component } from 'react';
import { connect } from 'react-redux';

import LoadingPage from './LoadingPage';
import PublicPostsList from './PublicPostsList';
import PopularLasttFilter from './filters/PopularLastFilter';
import HeaderSub from './header/HeaderSub';

import { startSetPostsSample } from '../actions/readings';

export class CategorizedReads extends Component {
  componentDidMount() {
    this.props.startSetPostsSample();
  }

  render() {
    return (
      <div>
        <HeaderSub category={this.props.category} />
        {this.props.isLoading ? (
          <LoadingPage />
        ) : (
          <div>
            <div className="content-container">
              <PopularLasttFilter relative />
            </div>
            <PublicPostsList grid="grid-home-first" category={this.props.category} range={14} />
          </div>
        )}
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startSetPostsSample: sampleSize => dispatch(startSetPostsSample(sampleSize))
});

const mapStateToProps = (state, props) => ({
  readings: state.readings.posts,
  isLoading: state.readings.isLoading,
  category: props.match.params.id.trim()
});

export default connect(mapStateToProps, mapDispatchToProps)(CategorizedReads);
