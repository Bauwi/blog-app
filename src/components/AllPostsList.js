import React, { Component } from 'react';
import { connect } from 'react-redux';

import { startSetPostsSample } from '../actions/posts';
import selectPosts from '../selectors/posts';

import AllPostsListItem from './AllPostsListItem';
import PostListFilters from './PostListFilters';

export class AllPostsList extends Component {
  componentWillMount() {
    this.props.startSetPostsSample(20);
  }

  renderListItem = () => {
    return this.props.posts.map(post => {
      return <AllPostsListItem {...post} />;
    });
  };

  render() {
    return (
      <div>
        <PostListFilters />
        <p>ALLpost List Here</p>
        {this.renderListItem()}
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startSetPostsSample: sampleSize => dispatch(startSetPostsSample(sampleSize))
});

const mapStateToProps = state => ({
  posts: selectPosts(state.posts, state.filters)
});

export default connect(mapStateToProps, mapDispatchToProps)(AllPostsList);
