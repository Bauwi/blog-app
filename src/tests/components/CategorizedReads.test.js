import React from 'react';
import { shallow } from 'enzyme';
import { CategorizedReads } from '../../components/CategorizedReads';
import run from '../fixtures/run';

const posts = run.posts.map(post => post.content);

test('should render CategorizedReads component properly', () => {
  const wrapper = shallow(<CategorizedReads
    readings={posts}
    category="music"
    isLoading={false}
    startSetPostsSample={() => {}}
  />);
  expect(wrapper).toMatchSnapshot();
});

test('should render CategorizedReads component properly on loading', () => {
  const wrapper = shallow(<CategorizedReads readings={posts} category="music" isLoading startSetPostsSample={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});

test('should call startSetPostsSample on mount', () => {
  const startSetPostsSample = jest.fn();
  const wrapper = shallow(<CategorizedReads
    readings={posts}
    category="music"
    isLoading={false}
    startSetPostsSample={startSetPostsSample}
  />);
  expect(startSetPostsSample).toHaveBeenCalled();
});
