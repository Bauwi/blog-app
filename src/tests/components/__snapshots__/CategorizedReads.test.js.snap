// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CategorizedReads component properly 1`] = `
<div>
  <HeaderSub
    category="music"
  />
  <div>
    <div
      className="content-container"
    >
      <Connect(PopularLasttFilter)
        relative={true}
      />
    </div>
    <Connect(PublicPostsList)
      category="music"
      grid="grid-home-first"
      range={14}
    />
  </div>
</div>
`;

exports[`should render CategorizedReads component properly on loading 1`] = `
<div>
  <HeaderSub
    category="music"
  />
  <Component />
</div>
`;
